css-ui-4/Overview.bs

```diff
@@ -1972,6 +1974,86 @@ Exclusion from Hit-testing: the 'pointer-events' property</h3>
 	with more possible values.
 	The effect of such values outside of SVG is currently not defined.
 
+<h3 id="inertness">
+Preventing All Interaction: the 'interactivity' property</h3>
+
+<pre class=propdef>
+Name: interactivity
+Value: auto | inert
+Initial: auto
+Applies to: all elements
+Inherited: yes
+Percentages: N/A
+Computed Value: as specified
+Animation type: discrete
+</pre>
+
+The 'interactivity' property specifies whether an element and its [=flat tree=] descendants
+(including [=text runs=])
+are [=inert=] or not.
+
+<dl dfn-type=value dfn-for=interactivity>
+	: <dfn>auto</dfn>
+	::
+		Unless otherwise specified by the host language,
+		the element is not [=inert=].
+
+	: <dfn>inert</dfn>
+	:: The element is [=inert=].
+</dl>
+
+<div algorithm>
+	When an [=element=] or [=text run=] is <dfn export>inert</dfn>:
+
+	* Hit-testing must act as if 'pointer-event' was ''pointer-events/none'',
+		regardless of its actual value.
+	* Text selection must act as if 'user-select' was ''user-select/none'',
+		regardless of its actual value.
+	* If the [=element=] or [=text run=] is editable,
+		it must behave as if it was non-editable.
+	* The [=element=] or [=text run=] should be ignored
+		for the purposes of find-in-page and similar searching actions.
+
+	User agents may allow the user 
+	to override the restrictions on find-in-page and text selection.
+</div>
+
+Note: [=Inert=] nodes generally cannot be focused, 
+and user agents do not expose the inert nodes 
+to accessibility APIs or assistive technologies.
+An inert subtree should not contain any content or controls 
+which are critical to understanding or using aspects of the page 
+which are not in the inert state. 
+Content in an inert subtree will not be perceivable by all users, or interactive. 
+Authors should not specify elements as inert 
+unless the content they represent are also visually obscured in some way. 
+In most cases, 
+authors should not specify the inert attribute on individual form controls. 
+In these instances, the <{input/disabled}> attribute is probably more appropriate.
+
+<div class=issue>
+	The HTML <{html-global/inert}> attribute is meant to be stronger
+	than the 'interactivity' property,
+	per CSSWG resolution.
```

Comment:
Hmm, do you know where is this resolution? In general:

 * I don't think that UA CSS works for shadow DOM (does `@scope` cross shadow boundaries).
 * It'd be nice to implement inert-escaping using this property (that's how it works in gecko already fwiw), and I don't see much reason for an author not to do the same?

But I guess I see the point of making `inert` take precedence... It just probably needs to be style system magic rather than expressed in terms of `@scope`.

Reply:
Rereading #10711, I think that "[inert] must win over CSS" might have been a carryover in my memory from when it was being proposed as a `visibility` value, to avoid existing code that sets `visibility` from starting to escape inertness accidentally. It looks like, aside from modal dialogs, we're actually okay with CSS defeating the inert attribute, so long as it's done by a new property that won't trigger problems in legacy code.

But also, you're right, `@scope` doesn't extend into shadow trees, and inertness needs to (particularly for the forced inertness from modal dialogs). So yeah, we'll need to track a bit coming from the host language.

I'm thinking:

1. The host language can indicate that a given element is "forced inert", which causes `interactivity: auto` to (behave as? compute to?) `inert`. This is set on the rest of the page when a modal dialog is active, for example.
2. We add a UA rule for `[inert] { interactivity: inert; }`, just relying on inheritance.
3. We add a UA rule (either using a normal selector, or a new pseudo-class if the qualities aren't exposed to selectors currently) for `dialog, etc { interactivity: auto; }`, so they'll escape inertness by default.

Reply:
Sorry I don't have the energy today to make a better thought through comment, but please consider tying inert-escaping behaviour to top layer. At least, don't hastily add some inert-escaping behaviour which will have all the potential downsides we were trying to avoid by not allowing it in the first place.

Reply:
> 1. The host language can indicate that a given element is "forced inert", which causes `interactivity: auto` to (behave as? compute to?) `inert`. This is set on the rest of the page when a modal dialog is active, for example.

From an implementer point of view, I'd prefer "behave as". Also, I think about it as the interactivity property affects the inertness in the host language, not that the host language affects the computed interactivity in CSS.

> 2. We add a UA rule for `[inert] { interactivity: inert; }`, just relying on inheritance.

:+1: 

> 3. We add a UA rule (either using a normal selector, or a new pseudo-class if the qualities aren't exposed to selectors currently) for `dialog, etc { interactivity: auto; }`, so they'll escape inertness by default.

I don't think we need this.` interactivity:auto` is the initial value, so this won't have an effect (unless you add !important). Also, the host language would just make sure dialog and its descendants don't have forced inertness?

Reply:
@lilles how wouldn't we need it? Note that the original comment about this not being an inherited property is no longer relevant.

Reply:
I think you still need the `dialog { interactivity: auto }` to make a modal `<dialog>` inside something like `<div inert>` work? I think it works everywhere now fwiw.

Reply:
That said... Do we really need the "force inert" concept? In Gecko, we currently implement this with an internal pseudo-class that applies to the top of inert subtrees ([here](https://searchfox.org/mozilla-central/rev/b765e2890b0eb85b24f54bc7ff04491fd0704e30/layout/style/res/ua.css#180-182)).

When a modal dialog is shown, or some element is fullscreen, we apply that pseudo-class to the root element, and another (`:-moz-topmost-modal`) to the dialog or fullscreen element.

I wonder if the solution is a single property authors can only set to `inert` (i.e. you can mark subtrees inert with CSS, but not escape inertness, that is handled by the UA)? I think that also addresses @alice's concerns, and doesn't require keeping two flags around.

Reply:
> I think you still need the `dialog { interactivity: auto }` to make a modal `<dialog>` inside something like `<div inert>` work? I think it works everywhere now fwiw.

You mean that the dialog should not be inert in the two cases below? Regardless of whether it's in the top layer or not?

```html
<div inert>
  <dialog></dialog>
</div>
<div style="interactivity:inert">
  <dialog></dialog>
</div>
```

Reply:
Being able to escape inertness was a requested ability https://github.com/w3c/csswg-drafts/issues/10711#issuecomment-2379610168 .  Is it possible that for dialog or other top layer blocking UI we add a `!important` inert style to prevent escaping it in the html content but otherwise allow the inert attribute to be escaped?

Reply:
@emilio's suggestion to only allow inert-escaping by the UA would certainly be the most conservative option and would indeed address my concerns.

My main concern is that allowing inert-escaping behaviour essentially based on specificity means that it would be possible to accidentally end up with elements/subtrees which escape inert without realising it, since there is intrinsically no visual impact of making something inert or un-inert (and the un-inerted element may actually be visually inaccessible). 

My suggestion to tie un-inerting to the top layer is based on the principle that AT behaviour follows from visual behaviour. Something being in the top layer is a reasonably reliable heuristic that it is visually obscuring other elements, particularly if it has a backdrop.

Working backwards through @scottaohara's very helpful example list of cases for un-inerting from the comment linked above:

(3) Elements outside of a modally-showing dialog which nevertheless need to be showing, in particular using popover elements which exist outside of the dialog. Frankly, this seems like a bug in popover to me, and could be addressed via Emilio's suggestion of having the UA manage un-inerting. In general, if something is rendered in the top layer, it probably _shouldn't_ be inert unless explicitly made inert.

(2) Live regions for global announcer systems. I'm incredibly sympathetic to this case, but as Scott notes, the ARIA notification proposal is a better fit for the cases not included in (3), and seems to be making good progress. Also, if we make top layer elements automatically un-inert when the page is inerted because of another top layer element, perhaps this could make use of that to promote the announcer element into the top layer and thus out of inert?

(1) UI tours, where (I think) both the popover explaining the relevant part of the UI, and the relevant part of the UI, should be un-inert and un-obscured. I'll admit, this is potentially the most compelling counter-example to my logic. That said, how is the "page is mostly dimmed" behaviour implemented? Could the same mechanism be used to implement "page is mostly inert" using the new property?

Reply:
> In general, if something is rendered in the top layer, it probably shouldn't be inert unless explicitly made inert.

I agree. I think UA rule to uninert may work for this case? E.g.

```css
/* TODO: match other things that may be in the top layer. */
dialog:open {
  interactivity: auto; /* Not !important so an author rule on the dialog can override this. */
}

[inert] {
  interactivity: inert;
}
```

Reply:
There's been more in this thread, but specifically:

> interactivity:auto is the initial value, so this won't have an effect (unless you add !important)

No, putting a rule into the cascade, even if it matches the initial value, causes it to defeat inheritance, as we only use the inherited value if *nothing* wins the cascade.

Reply:
> (3) Elements outside of a modally-showing dialog which nevertheless need to be showing, in particular using popover elements which exist outside of the dialog. Frankly, this seems like a bug in popover (or even dialog) to me, and could be addressed via Emilio's suggestion of having the UA manage un-inerting. In general, if something is rendered in the top layer, it probably _shouldn't_ be inert unless explicitly made inert.

This is a good point, which I've heard a few times. I opened this issue https://github.com/whatwg/html/issues/10811 to discuss this problem more generally. It likely should be "fixed" in WHATWG and not in CSSWG, since the top layer stack is there?

Reply:
`dialog:modal` would be more accurate than `dialog:open` in the UA rule above fwiw (in terms of matching the current behavior).